perun/collect/trace/strategy.py

```diff
@@ -28,213 +28,287 @@ class _Status(IntEnum):
-    lower priority and can be overwritten by the user supplied rules.
-
-    :param list extracted: list of extracted probes represented as a dictionaries
-    :param list custom: list of user-specified probes represented also as a dictionaries
-    :return list: the resulting list of merged probe locations
+            basic[k] = v
+
+    # Process the basic probes - find endings and compare them
+    beginnings, endings = dict(), dict()
+    for k, v in basic.items():
+        # Classify the probe as starting or ending according to the known suffixes
+        if not _check_suffix(v, suffix_pairs.values(), endings):
+            if not _check_suffix(v, suffix_pairs.keys(), beginnings):
+                # Not starting nor ending suffix, treat as regular probe name
+                beginnings.setdefault(v['name'], {'sample': v['sample'], 'suffix': []})
+    return delimited, beginnings, endings
+
+
+def _check_suffix(probe, suffix_group, cls):
+    """Checks if the probe name contains any suffix from the suffix group and if yes, classifies it
+    into the specified cls. Also stores the delimiter between the probe name and suffix, if any.
+
+    :param dict probe: the given probe specification
+    :param collection suffix_group: the collection of suffixes to check
+    :param dict cls: the resulting class
+    """
+    default = {'sample': probe['sample'], 'suffix': []}
+    # Iterate the suffixes from suffix group and check if the probe name ends with one of them
+    for suffix in suffix_group:
+        if probe['name'].lower().endswith(suffix):
+            name = probe['name'][:-len(suffix)]
+            # Store the delimiter if any
+            delimiter = ''
+            while name[-1] in _SUFFIX_DELIMITERS:
+                delimiter = name[-1] + delimiter
+                name = name[:-1]
+            cls.setdefault(name, default)['suffix'].append((suffix, delimiter))
+            return True
+    return False
+
+
+def _pair_suffixes(b_suffixes, e_suffixes, pairing):
+    """Pairs the starting and ending suffixes of one probe according to the provided
+    pairing function. Successfully paired suffixes are removed from the suffix lists.
+
+    :param list b_suffixes: the list of beginning suffixes associated with the probe
+    :param list e_suffixes: the list of ending suffixes
+    :param function pairing: function that takes beginning suffix (in lowercase) and should
+                             return the expected ending suffix or collection of them
+    :return list: the list of combined suffix pairs
     """
-    # Custom has higher priority than extracted
-    seen = set()
-    for probe in custom:
-        seen.add(probe['name'])
-    for probe in extracted:
-        if probe['name'] not in seen:
-            custom.append(probe)
-    return custom
+    # Check if the probe has matching starting and ending suffixes
+    pairs = []
+    for suffix in list(b_suffixes):
```

Comment:
Fix: I think it would be better to extract suffix and e_suffix. This will remove the ugly [0] and [1].